[Arrow] Properly use the parent's `array.offset` in many places in the scan

Tishj opened the pull request:
This PR fixes #5547 

It seems like the arrow arrays hit a certain chunk size limit, and are split up over multiple arrays while sharing the same buffer.
These offsets are put onto the struct array and should be forwarded to the child arrays, we did not do this previously.
(Or at least not consistently)

Tishj committed:
use parent offsets when scanning lists and varchars

Tishj committed:
add test for list of lists

Tishj committed:
fixed size list

Tishj committed:
binary and large_binary

Tishj committed:
add test with blobs

Tishj committed:
hunt for issues in struct/list combinations - fixed up some obvious issues looking at the patterns of the previous fixes

Tishj committed:
test conversion of time values, with all units

Tishj committed:
fix timestamptz conversion

Tishj committed:
fix duration

Tishj committed:
fix month_day_nanos

Tishj committed:
realize interval[months] can not be created from pyarrow..

Tishj committed:
fix decimals (16,32,64 and 128 bits)

Tishj committed:
use the ArrowBufferData abstraction

Tishj committed:
undo previous commit, order of operations between the cast and the addition mather | abstract the offset away into GetEffectiveOffset

Tishj committed:
add missing coverage for bools

Tishj committed:
add missing coverage for date32 and date64

Tishj committed:
add coverage for dictionaries

Tishj committed:
remove debug breakpoint code

Tishj committed:
rename test, it was misleading/wrong

Mytherin reviewed (approved):
Thanks! Looks great. Could you just look at the failing CI?

Tishj committed:
fixed a FIXME, turns out it does need parent_offset

Tishj committed:
use parent offset when setting the validity mask

Tishj committed:
add tests for null

Tishj commented:
> Thanks! Looks great. Could you just look at the failing CI?

Weirdly that's one of the tests I added, seems I didn't run it - and it was broken
I've fixed the test and added tests for nulls

Discovered another bug concerning dictionaries, structs and nulls - will likely fix this in a separate PR

Tishj committed:
need to initialize the offset in this test, because we now read it

Tishj committed:
fix enums in structs that are null

pdet commented on src/function/table/arrow_conversion.cpp:
```diff
@@ -18,24 +18,34 @@ static void ShiftRight(unsigned char *ar, int size, int shift) {
 	}
 }
 
+idx_t GetEffectiveOffset(ArrowArray &array, int64_t parent_offset, ArrowScanLocalState &state,
+                         int64_t nested_offset = -1) {
+	if (nested_offset != -1) {
+		// The parent of this array is a list
+		// We just ignore the parent offset, it's already applied to the list
+		return array.offset + nested_offset;
+	}
+	// Parent offset is set in the case of a struct, it applies to all child arrays
+	// 'chunk_offset' is how much of the chunk we've already scanned, in case the chunk size exceeds
+	// STANDARD_VECTOR_SIZE
+	return array.offset + parent_offset + state.chunk_offset;
+}
+
```
Should we have an offset struct that holds all the offset information and have this method in?
Then we can localize this offset logic and not pass two values all over.

Tishj replied:
We would still be populating this on the fly, the nested offset/parent offset is provided by the parent array.

I was thinking of writing this offset into the `ArrowArrayScanState` every time we scan a new chunk, so we don't repeat this logic everywhere, but the only thing we can get rid of that way is the parent_offset/nested offset if/else.

We would still need to keep the context of what type of offset this is though, because the `state.chunk_offset` would still need to be applied for every scanned Vector.

Tishj committed:
Merge remote-tracking branch 'upstream/main' into arrow_parent_offsets

Tishj commented:
Had this weird CI failure locally:
```
          for path in extension_paths_found:
              print(path)
              if path.endswith(extension_name + ".duckdb_extension"):
                  conn = duckdb.connect(db_name, config={'allow_unsigned_extensions': 'true'})
  >               conn.execute(f"LOAD '{path}'")
  E               duckdb.duckdb.InvalidInputException: Invalid Input Error: Extension "/project/httpfs/httpfs.duckdb_extension" version (v..1-dev) does not match DuckDB version (v0.0.0)
```

Probably caused by this being the version of the installed package:

```
      + pip install /tmp/cibuildwheel/repaired_wheel/duckdb-0.1.dev34218-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
  Processing /tmp/cibuildwheel/repaired_wheel/duckdb-0.1.dev34218-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
  Installing collected packages: duckdb
  Successfully installed duckdb-0.1.dev34218
```

Mytherin commented:
Thanks!
